src/meta/src/stream/scale.rs

```diff
@@ -2247,3 +2587,214 @@ impl GlobalStreamManager {
         (join_handle, shutdown_tx)
     }
 }
+
+// We redistribute parallel units (which will be ensembles in the future) through a simple consistent hashing ring.
+// Note that we have added some simple logic here to ensure the consistency of the ratio between each slot,
+// especially when equal division is needed.
+fn rebalance_units(
+    slots: BTreeMap<WorkerId, BTreeSet<ParallelUnitId>>,
+    total_unit_size: usize,
+) -> MetaResult<HashMap<WorkerId, BTreeSet<ParallelUnitId>>> {
+    let mut ch = ConsistentHashRing::new(1024);
+
+    for (worker_id, parallel_unit_ids) in &slots {
+        ch.add_worker(*worker_id, parallel_unit_ids.len() as u32);
+    }
+
+    let target_distribution = ch.distribute_tasks(total_unit_size as u32)?;
+
+    Ok(slots
+        .into_iter()
+        .map(|(worker_id, parallel_unit_ids)| {
+            (
+                worker_id,
+                parallel_unit_ids
+                    .into_iter()
+                    .take(
+                        target_distribution
+                            .get(&worker_id)
+                            .cloned()
+                            .unwrap_or_default() as usize,
+                    )
+                    .collect::<BTreeSet<_>>(),
+            )
+        })
+        .collect())
+}
+
+struct ConsistentHashRing {
```

Comment:
Since it's not compatible with the schedule algorithm in creation and replacing of steaming jobs, it will cause an unnecessary rebalance if the new created job is not placed in a weight-balanced manner. I think the consistent hash schedule algorithm introduced here is more reasonable and should be adapted in creation process.